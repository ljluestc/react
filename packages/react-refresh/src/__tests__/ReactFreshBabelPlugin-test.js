/** * Copyright (c) Meta Platforms, Inc. and affiliates. * * This source code is licensed under the MIT license found in the * LICENSE file in the root directory of this source tree. */const freshPlbabel.transform(input, {  babelrc: false,  configFile: false,  plugins: ['@babel/syntax-jsx','@babel/syntax-dynamic-import',[  freshPlugin,  {skipEnvCheck:  options.skipEnvCheck === undefined ? true : options.skipEnvCheck,// To simplify debugging tests:emitFullSignatures: true,...options.freshOptions,  },],...(options.plugins || []),  ],}).code,  );}describe('ReactFreshBabelPlugin', () => {  it('registers top-level function declarations', () => {// Hello and Bar should be registered, handleClick shouldn't.expect(  transform(`function Hello() { }function Bar() {  return <Hello />;}`),).toMatchSnapshot();  });  it('registers top-level exported function declarations', () => {expect(  transform(`e  return <h1 onClick={handleClick}>Hi</h1>;}export default function Bar() {  return <Hello />;}function Baz() {  return <h1>OK</h1>;}consexport const Bad = 42;`),).toMatchSnapshot();  });  it('registers top-level exported named arrow functions', () => {expect(  transform(`e  return <h1 onClick={handleClick}>Hi</h1>;};export let Bar = (props) => <Hello />;export default () => {urn <Hello />;};`  });  it('uses original function declaration if it get reassigned', () => {// This should register texpect(  transform(`function Hello() {  return <h1>Hi</h1>;}`),).toMatchSnapshot();  });  it('only registers pascal case functions', () => {// Should not get registered.expect(  transform(`fu}).toMatchSnapshot();  });  it('registers top-level variable declarations with function expressions', () => {// Hello and Bar should be registered; handleClick, sum, Baz, and Qux shouldn't.expect(  transform(`let Hello = function() {  function handleClick() {}  retconst Bar = function Baz() {  return <Hello />;};function sum() {}let Baz = 10;var Qux;`),).toMatchSnapshot();  });  it('registers top-level variable declarations with arrow functions', () => {// Hello, Bar, and Baz should be registered; handleClick and sum shouldn't.expect(  tra  const handleClick = () => {};  return <h1 onClick={handleClick}>Hi</h1>;}const Bar = () => {  return <Hello />;};var Baz = () => <div />;var sum = () => {};`),).toM// TODO: we might want to handle HOCs at usage site, however.// TODO: it would be nice if we could always avoid registering// a function that is known to return a function or other non-node.expect(  transform(`let connect = () => {  function Comp() {const handleClick = () => {};return <h1 onClick={handleClick}>Hi</h1>;  }  return Comp;};function withRouter() {  return function Child() {const handleClick = () => {};return <h1 onClick={handleClick}>Hi</h1>;  }`),).toMatchSnapshot();  });  it('ignores complex definitions', () => {expect(  transform(`let A = foo ? () => {  return <h1>Hi</h1>;} : nullconst B = (function Foo() {  return <h1>Hi</h1>;})();let C = () => () => {  return <h1>Hi</h1>;};let D = bar && (() => {  ret`),).toMatchSnapshot();  });  it('ignores unnamed function declarations', () => {expect(  transform(`export default function() {}`),).toMatchSnapshot();  });  it('registers likely HOCs with inline functions', () => {expect(  transform(`const A = forwardRef(function() {  return <h1>Foo</h1>;});const B = memo(React.forwardRef(() => {  return <h1>Foo</h1>;}));export default React.memo(forwardRef((props, ref) => {  return <h1>Foo</h1>;}));`),).toM  transform(`export default React.memo(forwardRef(function (props, ref) {  return <h1>Foo</h1>;}));`),).toMatchSnapshot();expect(  transform(`export default React.memo(forwardRef(function Named(props, ref) {  return <h1>Foo</h1>;}));`),).toMatchSnapshot();  });  it('ignores higher-order functions that are not HOCs', () => {expect(  transform(`const throttledAlert = throttle(function() {  aleconst TooComplex = (function() { return hello })(() => {});if (cond) {  const Foo = thing(() => {});}`),).toMatchSnapshot();  });// Foo, Header, and B get registered.// A doesn't get registered because it's not declared locally.// Alias doesn't get registered because its definition is just an identifier.expect(  transform(`import A from './A';import Store from './Store';Store.subscribe();const Header = styled.div\`color: red\`const StyledFactory1 = styled('div')\`color: hotpink\`const StyledFactory2 = styled('div')({ color: 'hotpink' })const StyledFactory3 = styled(A)({ color: 'hotpink' })const FunnyFactory = funny.factory\`\`;let Alias1 = A;let Alias2 = A.Foo;const Dict = {};function Foo() {  return (<div><A /><B /><StyledFactory1 /><StyledFactory2 /><StyledFactory3 /><Alias1 /><Alias2 /><Header /><Dict.X /></div>  );}const B = hoc(A);// This is currently registered as a false positive:const NotAComponent = wow(A);// We could avoid it but it also doesn't hurt.`),).toMatchSnapshot();  });  it('registers identifiers used in React.createElement at definition site', () => {// When in doubt, register variables that were used in JSX.// Foo, Header, and B get registered.// A doesn't get registered because it's not declared locally.// Alias doesn't get registered because its definition is just an identifier.expect(  traimport Store from './Store';Store.subscribe();const Header = styled.div\`color: red\`const StyledFactory1 = styled('div')\`color: hotpink\`const StyledFactory2 = styled('div')({ color: 'hotpink' })const StyledFactory3 = styled(A)({ color: 'hotpink' })const FunnyFactory = funny.factory\`\`;let Alias1 = A;let Alias2 = A.Foo;const Dict = {};function Foo() {  return [React.createElement(A),React.createElement(B),React.createElement(StyledFactory1),React.createElement(StyledFactory2),ReactReact.createElement(Alias2),jsx(Header),React.createElement(Dict.X),  ];}React.createContext(Store);const B = hoc(A);// This is currently registered as a false positive:const NotAComponent = wow(A);// We could avoid it but it also doesn't hurt.`),  });  it('registe  transform(`function Foo() {  return <h1>Hi</h1>;}export default hoc(Foo);export const A = hoc(Foo);`),).toMatchSnapshot();  });  it('gener  transform(`export default function App() {  const [foo, setFoo] = useState(0);  React.useEffect(() => {}); `),).toMatchSnapshot();  });  it('generates signatures for function expressions calling hooks', () => {// Unlike __register__, we want to sign all functions -- not just top level.// This lets us support editing HOCs better.// For function declarations, __signature__ is called on next line.// Foexpect(  transform(`export const A = React.memo(React.forwardRef((props, ref) => {  const [foo, setFoo] = useState(0);  React.useEffect(() => {});  return <h1 ref={ref}>{foo}</h1>;}));export const B = React.memo(React.forwardRef(function(props, ref) {  const [foo, setFoo] = useState(0);  React.useEffect(() => {});}));function hoc()const [foo, setFoo] = useState(0);React.useEffect(() => {});return <h1 ref={ref}>{foo}</h1>;  };}export let C = hoc();).toMatchSnapshot();  });  it('includes custom hooks into the signatures', () => {expect(function useFancyState() {  const [foo, setFoo] = React.useState(0);  useFancyEffect();  return foo;}const useFancyEffect = () => {  React.useEffect(() => {});};export default function App() {  const bar = useFancyState();  return <h1>{bar}</h1>;}`),).toMatchSnapshot(); // but would fail for Babel  transform(`import {useFancyState} from './hooks';export default function App() {  const bar = useFancyState();  return <h1>{bar}</h1>;}`,  plugins: ['@babel/transform-modules-commonjs'],},  ),).toMatchSnapshot();  });  it('generates valid signature for exotic ways to call Hooks', () => {eimport FancyHook from 'fancy';export default function App() {  function useFancyState() {const [foo, setFoo] = React.useState(0);useFancyEffect();return foo;  }  const baz = FancyHook.useThing();  React.useState();  useThePlatform();  return <h1>{bar}{baz}</h1>;}`),).toMatchSnapshot();  });  it('does not consider require-like methods to be HOCs', () => {// None of these were declared in this file.// It's bad to register them because that would trigger// mo// they are used in JSX.expect(  transform(`const A = require('A');const B = foo ? require('X') : require('Y');const C = requireCond(gk, 'C');const D = import('D');export default function App() {  return (<div>  <A />  <B />  <C />  <D  );}`),).toMatchSnapshot();  })  transform(`export default () => useContext(X);export const Foo = () => useContext(X);module.exports = () => useContext(X);const Bar = () => useContext(X);const Baz = memo(() => useContext(X));c).toMatchSnapshot();  });  it('uses custom identifiers for $RefreshReg$ and $RefreshSig$', () => {expect(  trauseContext(X)return <Foo />  };`,{  freshOptions: {refreshReg: 'import.meta.refreshReg',refreshSig: 'import.meta.refreshSig',).toMatchSnapshot();  expect(() =>  transform(`export default function BabelEnv () { return null };`, {envName,skipEnvCheck: false,  }),).toThrowError(  'ReenvName +'". If you want to override this check, pass {skipEnvCheck: true} as plugin options.',);  });  it('does not get tripped by IIFEs', () => {expect(  transform(`while (item) {useFoo();  })(item);}  `),).toMatchSnapshot();  });  it('supports typescript namespace syntax', () => {expect(  transform(`namespace Foo {  expexport const B1 = B;  }  export const C = () => {};  export function D() {};  namespace NotExported {export const E = () => {};}  `,{plugins: [['@babel/plugin-syntax-typescript', {isTSX: true}]]},  ),).toMatchSnapshot();  });});