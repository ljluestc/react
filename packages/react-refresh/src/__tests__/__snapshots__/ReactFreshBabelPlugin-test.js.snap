// Jest Snapshot v1, https://goo.gl/fbAQLP_s2 = $RefreshSig$(),_s3 = $RefreshSig$(),_s4 = $RefreshSig$(),_s5 = $RefreshSig$(),_s6 = $RefreshSig$();export default _s(() => {  _s();  return useContext(X);}, "useContext{}");  _s2();  return useContext(X);};_s2(F  _s3();  return useContext(X);}, "useContext{}");const Bar = () => {  _s4();  return useContext(X);};_s4(Ba}, "useContext{}")), "useContext{}");_c4 = Baz;const Qux = () => {  $RefreshReg$(_c2, "Bar");$RefreshReg$(_c4, "Baz");$RefreshReg$(_c5, "Qux");`;exportconst C = requireCond(gk, 'C');const D = import('D  return <div>  <A />  <C />  }`;exports[  var _s = $RefreshSig$();  _s(item => {_s();use}`;exports[`ReactFreshBabelPlugin gene  _s();  c  return <h1>{foo}</h1>;}_s(App,var _s = $RefreshSig$(),_s  React.useEffect(() => {}, "useStaexport const B = _s2(React.  React.useEffect(() => {});  return <h1 ref={ref}>{foo}</h1>;}, "useState{[foo, setFoo](0)}\\nuseEffect{}")), "useState{[foo, setFoo](0)}\\nuseEffect{}")), "useState{[foo, setFoo](0)}\\nuseEffect{}");_c6 = B;function hoc() {  var _s3 = $RefreshSig$();  return _s3(function Inner() {_sreturn <h1 ref={ref}>{foo}</h1>;  }, "useState{[foo, se$RefreshReg$(_c2, "A$React.memo");$RefreshReg$(_c3, "A");$RefreshReg$(_c4, "B$React.memo$React.forwardRef");$RefreshReg$(_c5, "B$React.memo");$RefreshReg$(_c6, "B");`;exports[`ReactFreshBabelPlugin generates valid signature for exotic ways to call Hooks 1`] = `var _s2 = $RefreshSig$();import FancyHook from 'fancy';export default function App() {  _s2();  var _s = $RefreshSig$();  function useFancyState() {_s();const [foo, setFoo] = React.useState(0);useFancyEffect();return fo  const  useThePlatform();return [FancyHook.useThing];});_c = App;var _c;$RefreshReg$(_c, "App");`;exports[`ReactFreshBabel  function Comp() {const};function withRouter() {  return function Child() {const handleClick = () => {};return <h1 onClick={handleClick}>Hi</h1>;  `;exports[`ReactFreshBabelPlugin ignores complex definitions 1`] = `let A = foo ? () => {} : null;const B = function Foo() {  retur  return <h1>Hi</h1>;};let D = bar && (() => {  return <h1>Hi</h1>;}const throttledAlert = throttle(function () {});const }(() =>}`;_s2 = $RefreshSig$(),_s3 = $RefreshSig$();function useFancyState() {  _s();  const [foo,urn foo;}_s(use});const useFancyEffect = () => {  _s2();  React.useEffect(() => {});};_s2(useFancyEffect, "useEffect{}");export default function App() {  _s3();  const bar = useFancyState();  return <h1>{bar}</h1>;}_s3(App, "useFancyState{bar}", false, function () {  return`;exports[`ReactFreshBabelPlugin includes custom hooks into the signatures when commonjs target is used 1`] = `"use strict";Object.defineProperty(exports, "__esModule", {  value: true});exports.  return <h1>{bar}</h1>;}_s(App, "useFancyState{bar});_c = App;var _c;$RefreshReg$(_c, "App");`;expo  return 2 * 2;}`;exports[`ReactFreshBabelPlugin registers capitalized identifiers in HOC calls 1`] = `function Foo() { export default _c2 = _c3 = A;_c4 = B;var _c, _c2, _c3, _c4;$RefreshReg$(_c, "Foo");$RefreshReg$(_c2, "%default%");$RefreshReg$(_c3, "A");$RefreshReg$(_c4, "B");`;exports[`ReactFreshBabelPlugin registers identifiers used in JSX at definition site 1`] = `import A from './A';imconst Header = styled.div\`color: red\`;_c = Header;_c2 = StyledFactory1;const StyledFactory2 = styled('div')({  color:_c3 = StyledFactory2;  color: 'hotpink'});const FunnyFactory = funny.factory\`\`;let Alias1 = A;let Alias2 = A.Foo;con}_c5 = Foo;const NotAComponent = wow(A); // We could avoid it but it also doesn't hurt._c7 = NotAComponent;var _c, _c2, _c3, _c4, _c5, _c6, _c7;$RefreshReg$(_c, "Header");$RefreshReg$(_c2, "StyledFactory1");$RefreshReg$(_c3, "StyledFactory2");$RefreshReg$(_c4, "StyledFactory3");$RefreshReg$(_c5, "Foo");$`;exports[`ReactFreshBabelPlugin registers identifiers used in React.createElement at definition site 1`] = `import A from './A';impconst Heaconst Sconst StyledFactory2 = styled('div')({  _c3 = StyledFactory2;const StyledFactory3 = styled(A)({  color: 'hotpink'});const FunnyFactory = funny.factory\`\`;letconst Dict = {};function Foo() {  React.createContext(Store);const B = hoc(A); // This is currently registered as a false positive:_c6 = B;const NotAComponent = wow(A);$RefreshReg$(_c3, "StyledFactory2");$RefreshReg$(_c4, "StyledFactory3");$$RefreshReg$(_c7, "NotAComponent");`;  return <h1>Foo</h1>;});_c2 = A;const B =}));_c5 = B;expo}));var _c, _c2, _c3, _c4, _c5, _c6, _c7, _c8;$RefreshReg$(_c, "A$forwardRef");$RefreshReg$(_c2, "A");$R$RefreshReg$(_c5, "B");$RefreshReg$(_c6, "%default%$React.memo$forwardRef");$RefreshReg$(_c7, "%default%$React.memo");$Rexport default _c3 = React.memo(_c2 = forwardRef(_c = function (props, ref) {  return <h1>Foo</h1>;}));var _c, _c2, _c3;$RefreshReg$(_c, "%default%$React.memo$forwardRef");$Re`;exports[`ReactFreshBabelPlugin registers likely HOCs with inline functions 3`] = `export default _c3 = React.memo(_c2 = forwardRef(_c = function Named(props, ref) {  return <h1$RefreshReg$(_c2, "%default%$React.memo");$RefreshReg$(_c3, "%default%");`;exports[`ReactFreshBabelPlugin registers top-level exported function declarations 1`] = `ex}export default function Bar() {  return <Hello />;}_c2 = Bar;function Ba}_c3 = Baz;const Nexport function sum() {}export const Bad = 42;var _c, _c2, _c3;$RefreshReg$(_c, "Hello");$RefreshReg$(_c2, "Bar");$export const Hello = () => {  function handleClick() {}  return <h1 onClick={handleClick}>Hi</h1>;};export let Bar = props => <Hello />;_c2 = Ba  // This one should be ignored.  });var _c, _c2;$RefreshReg$(_c, "He`;exports[`ReactFreshBabelPlugin registers top-level function declarations 1`] = `function}_c = Hello;function Bar() {  return <Hello />;}`;exports[`ReactFreshBabelPlugin registers top-level variable declarations with arrow functions 1`] = `let Hello = () => {  c  return $Refres`;exports[`ReactFreshBabelPlugin registers top-level variable declarations with function expressions 1`] = `le};_c = Hello;const Bar = function Baz() {  return <Helvar Qux;var _c, _c2;$RefreshReg$(_c, "Hello");$RefreshReg$(_c2, "Bar");`;exports[`ReactFreshBabelPlugin supports typescript namespace syntax 1`] = `namespace Foo {export const A = () => {};function B() {};exportort const C = () => {};  _c3 = C;  namespace NotExported {exp}var _c, $Refres`;exports[`ReactFreshBabelPlugin uses custom identifiers for $RefreshReg$ and $RefreshSig$ 1`] = `va  return <Foo />;}_s(Bar, "useContext{}");_c = Bar;;var _c;import.meta.refreshReg(_c, "Bar");`;exports[`ReactFreshBabelPlugin uses original function declaration if it get reassigned 1`] = `fu}_c = Hello;Hello = connect(Hello);var _c;$RefreshReg$(_c, "Hello");`;